components/script_plugins/unrooted_must_root.rs

```diff
@@ -142,26 +145,103 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnrootedPass {
+impl<'a, 'b, 'tcx> UnrootedCx<'a, 'b, 'tcx> {
+    fn is_unrooted_ty(&self, ty: &ty::TyS, in_new_function: bool) -> bool {
+        let mut ret = false;
+        let cx = self.late_cx;
+        ty.maybe_walk(|t| {
+            match t.sty {
+                ty::TyAdt(did, _) => {
+                    if cx.tcx.has_attr(did.did, "must_root") {
+                        ret = true;
+                        false
+                    } else if cx.tcx.has_attr(did.did, "allow_unrooted_interior") {
+                        false
+                    } else if match_def_path(cx, did.did, &["core", "cell", "Ref"])
+                            || match_def_path(cx, did.did, &["core", "cell", "RefMut"])
+                            || match_def_path(cx, did.did, &["core", "slice", "Iter"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "Entry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "OccupiedEntry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "VacantEntry"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "map", "Iter"])
+                            || match_def_path(cx, did.did, &["std", "collections", "hash", "set", "Iter"]) {
+                        // Structures which are semantically similar to an &ptr.
+                        false
+                    } else if did.is_box() && in_new_function {
+                        // box in new() is okay
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyParam(param_ty) => {
+                    let ty_param_def = cx.tcx.generics_of(self.def_id).type_param(&param_ty, cx.tcx);
+                    // will work for `foo::<Foo>(..)`
+                    // but won't get enough information in `bar<T>(){foo::<T>(..);}`?
+                    if cx.tcx.has_attr(ty_param_def.def_id, "must_root") {
+                        ret = true;
+                        false
+                    } else {
+                        true
+                    }
+                },
+                ty::TyRef(..) => false, // don't recurse down &ptrs
+                ty::TyRawPtr(..) => false, // don't recurse down *ptrs
+                ty::TyFnDef(..) | ty::TyFnPtr(_) => false,
+                _ => true
+            }
+        });
+        ret
+    }
+}
+
+impl<'a, 'b, 'tcx> MirVisitor<'tcx> for UnrootedCx<'a, 'b, 'tcx> {
+    fn visit_local_decl(&mut self, _local: mir::Local, decl: &mir::LocalDecl<'tcx>) {
+        if decl.is_user_variable {
+            if self.is_unrooted_ty(decl.ty, self.in_new_function) {
+                self.late_cx.span_lint(UNROOTED_MUST_ROOT, decl.source_info.span, "Type of binding/expression must be rooted.")
             }
         }
+        // how to check return value and arguments?
```

Comment:
the return value and arguments are locals -- so you could just check the number, yes. As for passing `mir`, I would put a `&Mir<'tcx>` in the `UnrootedCx`.

Reply:
Your comment above is not displayed in files changed section, that's why I have commented another dot below.